language/vm/src/abi.rs

```diff
@@ -0,0 +1,302 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    access::ModuleAccess,
+    file_format::{
+        CompiledModule, FieldDefinition, FunctionHandle, Kind, SignatureToken, StructDefinition,
+        StructFieldInformation, TypeParameterIndex,
+    },
+};
+use move_core_types::{identifier::Identifier, language_storage::ModuleId};
+
+/// Defines the Application Binary Interface (ABI) for Move modules, the `check_compatibility`
+/// function for checking ABI compatibility of two modules, and the `can_update` function for
+/// determining whether a new version of a module can safely update the old one.
+
+/// A normalized version of `SignatureToken`, a type expression appearing in struct or function
+/// declarations. Unlike `SignatureToken`s, `TypeABI`s from different modules can safely be compared.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum TypeABI {
+    Bool,
+    U8,
+    U64,
+    U128,
+    Address,
+    Signer,
+    Struct {
+        id: ModuleId,
+        name: Identifier,
+        type_arguments: Vec<TypeABI>,
+    },
+    Vector(Box<TypeABI>),
+    TypeParameter(TypeParameterIndex),
+    Reference(Box<TypeABI>),
+    MutableReference(Box<TypeABI>),
+}
+
+/// ABI for a `FieldDefinition`. The `name` is included even though it is
+/// metadata that it is ignored by the VM. The reason: names are important to clients. We would
+/// want a change from `Account { bal: u64, seq: u64 }` to `Account { seq: u64, bal: u64 }` to be
+/// marked as incompatible. Not safe to compare without an enclosing `StructABI`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FieldABI {
+    name: Identifier,
+    type_: TypeABI,
+}
+
+/// ABI for a struct declaration. Not safe to compare without an associated `ModuleId` or `ModuleABI`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct StructABI {
+    name: Identifier,
```

Comment:
Make the fields `pub`?

Note: I've come to think that hiding fields under getters in plain data objects (without particular internal invariants) is not so useful and in fact slightly detrimental in Rust. The builder pattern is arguably a lifesaver in Java where interior mutability is the default but that's not the case in Rust.